wikidict/parse.py

```diff
@@ -0,0 +1,135 @@
+
+from typing import (
+    TYPE_CHECKING,
+    Dict,
+    Generator,
+    List,
+    Tuple,
+)
+
+if TYPE_CHECKING:  # pragma: nocover
+    from xml.etree.ElementTree import Element
+
+import wikitextparser as wtp
+import wikitextparser._spans
+
+# As stated in wikitextparser._spans.parse_to_spans_innerloop():
+#   If the byte_array passed to parse_to_spans contains n WikiLinks, then
+#   this function will be called n + 1 times. One time for the whole byte_array
+#   and n times for each of the n WikiLinks.
+#
+# We do not care about links, let's speed-up the all process by skipping the n times call.
+# Doing that is a ~30% optimization.
+wikitextparser._spans.WIKILINK_PARAM_FINDITER = lambda *_: ()
+
+
+Sections = Dict[str, List[wtp.Section]]
+
+
+def xml_iter_parse(file: str) -> Generator["Element", None, None]:
+    """Efficient XML parsing for big files.
+    Elements are yielded when they meet the "page" tag.
+    """
+    import xml.etree.ElementTree as etree
+
+    doc = etree.iterparse(file, events=("start", "end"))
+    _, root = next(doc)
+
+    start_tag = None
+
+    for event, element in doc:
+        if (
+            start_tag is None
+            and event == "start"
+            and element.tag == "{http://www.mediawiki.org/xml/export-0.10/}page"
+        ):
+            start_tag = element.tag
+        elif start_tag is not None and event == "end" and element.tag == start_tag:
+            yield element
+            start_tag = None
+
+            # Keep memory low
+            root.clear()
+
+
+def xml_parse_element(element: "Element") -> Tuple[str, str]:
+    """Parse the *element* to retrieve the word and its definitions."""
+    revision = element[3]
+    if revision.tag == "{http://www.mediawiki.org/xml/export-0.10/}restrictions":
+        # When a word is "restricted", then the revision comes just after
+        revision = element[4]
+    elif not revision:
+        # This is a "redirect" page, not interesting.
+        return "", ""
+
+    # The Wikicode can be at different indexes, but not ones lower than 5
+    for info in revision[5:]:
+        if info.tag == "{http://www.mediawiki.org/xml/export-0.10/}text":
+            code = info.text or ""
+            break
+    else:
+        # No Wikicode, maybe an unfinished page.
+        return "", ""
+
+    word = element[0].text or ""  # title
+    return word, code
+
+
+def process(file: Path, locale: str) -> Dict[str, str]:
+    """Process the big XML file and retain only information we are interested in."""
+    words: Dict[str, str] = defaultdict(str)
+
+    print(f">>> Processing {file} ...", flush=True)
+    for element in xml_iter_parse(str(file)):
+        word, code = xml_parse_element(element)
+        if not word or not code or ":" in word:
+            continue
+
+        if any(head_section in code for head_section in head_sections[locale]):
+            words[word] = code
+        else:
+            print(f"Ignored [{word}]")
```

Comment:
```suggestion
            print(f"Ignored [{word}]", flush=True)
```

Reply:
that's a debug statement, I'll remove it